### ISSUE TITLE
crashes

### ISSUE THREAD

**Issue by rofl0r** [glouw/minimidi#1]
with this file https://github.com/oxesoft/oxefmsynth/blob/master/src/converter/testtune.mid (i get a bus error)

with other files (e.g.  https://github.com/Mindwerks/wildmidi/files/5489862/test.mid.zip) i get barely audible and noisy playback and almost 100% cpu load. somehow the stuff you do with the thread doesn't seem optimal.
maybe you can get inspiration from https://github.com/rofl0r/c-flod/blob/master/backends/sdlwriter.c

**Comment by glouw** [comment]
Thanks. The high CPU usage is likely coming from the implementation of SDL_Atomic Set and Get. I do not know too much about them, and I do regret using them. I plan to rewrite an even more compact version of minimidi now that I have a strong understanding of the midi internals. As for the bus error, I'll take a look later this week.

I will keep you posted

**Comment by rofl0r** [comment]
afaik SDL already spawns a thread for the audio playback, thus the callback dance in my sdlwriter backend. i personally found it much nicer to work with the synchronous API of libao, but YMMV.

**Comment by glouw** [comment]
That's right, the proper fix would be to get rid of the consumer thread. I can probably turn down the sample rate as well. I do not think sine waves need 44.1K sample rate.

**Comment by glouw** [comment]
Not sure if you're still interested, but I removed the atomic sets / gets. Performance should be desirable. As for certain midi compatibility, this will be forever ongoing. I'm currently in the process of mapping some of my favorite tunes to the instrument list

**Comment by rofl0r** [comment]
nice! btw, i realized after filing this here that the file which threw the bus error had actually been overwritten with the textual output of minimidi, so the file format was quite unlike midi and there are probably some sanity checks missing. so i seemed to have fuzzed your implementation unwillingly...

i will try out your improvements asap.

**Comment by glouw** [comment]
Cool, yeah. I guess minimidi works well with general midi (https://en.wikipedia.org/wiki/General_MIDI) which is supported really well when exporting Guitar Pro files (found on ultimate-guitar.com) as midi with tuxguitar: https://www.youtube.com/watch?v=CVJd5w0Kk00.

I'm currently working on a handful of visualizers, and fleshing out the cohesiveness of the instruments. Making various waves go well together is really hard, so I've learned. I may dive into a bit of OPL with fm/am modulation, but as is basic 8 bit waves sound pretty nostalgic, and I'm really enjoying it.
